### PR TITLE
Rude AT Command Support

### PR DESCRIPTION
**Pull request by stieg**
This change is built on top of PR #664, so please review that first prior to reviewing this.

This change adds the ability to support "Rude AT commands".  Rude at commands are AT commands that show up during the middle of another command instead of being queued.  They are considered rude because they interrupt the other command (and can even ruin it).  But sometimes you have to work with the cards you're dealt and this is one of those times.

In brief testing it seems to help the WiFi driver cope with a flood of messages, but there is still the chance that if you flood it with messages it simply will not be able to handle them or may even freeze.  I'm doing my best to prevent that as is reasonable.

### CODE REVIEW

File: src/drivers/esp8266_drv.c [hunk 1]
```diff
@@ -1128,38 +1206,163 @@ bool esp8266_drv_init(struct Serial *s, const int priority,
...
-                                   const char* dst_ip,
-                                   const unsigned int dst_port)
+                                   const char* addr,
+                                   const unsigned int port)
 {
-        const int chan_id = get_next_free_channel_num();
-        if (chan_id < 0) {
+        /* This is synchronous code for now */
+        struct channel_sync_op *cso = &esp8266_state.comm.connect_op;
+        xSemaphoreTake(cso->op_semaphore, portMAX_DELAY);
```
Comment (brentpicasso) [inline src/drivers/esp8266_drv.c]: what if we can never get the semaphore? We should bail out or fail somehow. Otherwise the thing could get wedged.

Comment (brentpicasso) [inline src/drivers/esp8266_drv.c]: I realized the PR notes and this builds on the previous one I reviewed. holding off comments.

**Comment by stieg** [comment]
Ok @brentpicasso this is ready for your eyes :)

File: src/modem/at.c [hunk 2]
```diff
@@ -253,7 +253,20 @@ static void at_task_run_bytes_read(struct at_info *ati, char *msg)
          * Our rx state will dictate how we process this message beacuse that
          * allows us to know previous messages and what message type to expect.
          */
-        switch (ati->rx_state) {
+        //enum at_rx_state state = ati->rx_state;
```
Comment (brentpicasso) [inline src/modem/at.c]: should we remove commented code?

**Comment by brentpicasso** [comment]
looks great, aside from one question, optionally address. otherwise :ship: !

File: src/modem/at.c (hunk 2 above)
Comment (stieg) [inline src/modem/at.c]: Can do.

Comment (stieg) [inline src/modem/at.c]: And hell you just saved my butt.  That was some code that I left in there for testing that should have been removed.  Thanks :)

**Comment by stieg** [comment]
Ok... turns out I got ahead of myself with this one.  Need to do more.  So closing this until its really ready :\

**Pull request closed without merging**